### PR TITLE
openstreetmap-carto project file for TagInfo

### PR DESCRIPTION
**Pull request by ImreSamu**
**STATUS (2018May08) :  As I understand and accept that the maintainence problem is not optimal for my "toy code", so I will move my code to the separate repo/project, and I will close this pull requests in the next days.**

----  

background:   #961 

This is a minimalistic approach: only the osm keys parsed and reported.

I have created a separated directory `./taginfo-project/`

New files:
* `./taginfo-project/README.md`    : basic readme
* `./taginfo-project/generate-taginfo-project-file.py` 
* `./taginfo-project/taginfo-openstreetmap-carto.json`   -> this is the result , for the taginfo! 

Status:
- [x] First working version 
- [x] add ARRAY parsing ( `tags @> ARRAY['building',...] ` )
- [x] the result imported to the Taginfo dev system, and everything is OK.
(https://github.com/gravitystorm/openstreetmap-carto/blob/master/CONTRIBUTING.md#sql-style-guidelines)
- [x] integrate to the workflow ( .travis ? )

### CODE REVIEW

**Comment by kocio-pl** [comment]
> integrate with the CONTRIBUTING.md#sql-style-guidelines

I'm against adding anything to do by hand, especially if it could be easily automated. I still don't have the idea how to make it efficiently after each release, though. I think that while we can include this script, as it is closely related to our code, but running this tool is another story.

**Comment by ImreSamu** [comment]
@kocio-pl : 
>integrate with the CONTRIBUTING.md

I need to re-think,  but you can expect some extra documentation line ...


# Testing 
 

temporary "taginfo-project"   test

* http://as-af.taginfo-dev.opengeodata.hu/projects/osmcartov2#overview
* http://as-af.taginfo-dev.opengeodata.hu/projects/osmcartov2#tags

please help:
* the "Objects" detections is correct? ( node,way,area, realation)
* Now this is the text - but I am not satisfied, what is your suggestions?
     * `Used in the osm2pgsql database backend, see more in the github repo`
    * `Used as a hstore tags-> in the database backend, see more in the github repo`
* Please test the "More->" part  (  github search )

**Comment by kocio-pl** [comment]
> I need to re-think, but you can expect some extra documentation line ...

Sorry, I've made a mistake. It's OK to make the style more strict and consistent (it might be even preferable). I meant RELEASES.md, which you didn't even mentioned... 

Still it means that this script should be triggered somehow and the best moment seems to be release time.

File: taginfo-project/generate-taginfo-project-file.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
+#!/usr/bin/env python3
+
+# -------------------------------------------------------------------------------
+#
+# This code generate a taginfo project list file (see more https://wiki.openstreetmap.org/wiki/Taginfo/Projects )
```
Comment (kocio-pl) [inline taginfo-project/generate-taginfo-project-file.py]: generate->generates

**Comment by kocio-pl** [comment]
I think JSON file should not be included in the final version - it will be automatically generated only when needed. It should be always listed in `.gitignore` file.

**Comment by ImreSamu** [comment]
> I think JSON file should not be included in the final version -

I don't think I've understood all that,  
Could you clarify why don't want the final JSON file? ( = taginfo project file )

According to the ["Taginfo/Projects/Integration into Taginfo"](https://wiki.openstreetmap.org/wiki/Taginfo/Projects)
_"Taginfo will periodically (usually daily) poll the project file and add the information to its own database."_

And I have planned the project file should be here : 
*  https://raw.githubusercontent.com/gravitystorm/openstreetmap-carto/master/taginfo-project/taginfo-openstreetmap-carto.json   

And I have not added a timestamp inside the JSON,  so my `git status` don't detect as a change if the content is the same.

File: taginfo-project/generate-taginfo-project-file.py (hunk 1 above)
Comment (ImreSamu) [reply taginfo-project/generate-taginfo-project-file.py]: Thank you,  fixed!

**Comment by kocio-pl** [comment]
I guess this file will be created by Travis on next merge, isn't it?

**Comment by kocio-pl** [comment]
The script and documentation should be placed inside the `scripts` directory, just like [lua](https://github.com/gravitystorm/openstreetmap-carto/tree/master/scripts/lua).

**Comment by ImreSamu** [comment]
>I guess this file will be created by Travis on next merge, isn't it?

I am using the Travis only for checking.
* testing:  my code is running? 
* and a simple `git diff` - what has been changed in the taginfo project file. 

if you not need, you can remove.

imho:  some human checking is useful because my code is not _well_ tested.  ( code parsing  is hard )





>The script and documentation should be placed inside the scripts 

ok, no problem.

like this?

old files                                          |  new files
---------------------------------------------------| ------------------
./taginfo-project/generate-taginfo-project-file.py | ./scripts/taginfo-project/generate-taginfo-project-file.py 
./taginfo-project/README.md                        | ./scripts/taginfo-project/README.md


imho:  than better if we move the taginfo JSON  project file to the root.  ( so no need a special folder ) 

my suggestion:

old project file                                   |  new project file 
---------------------------------------------------| ------------------
./taginfo-project/taginfo-openstreetmap-carto.json | ./taginfo-openstreetmap-carto.json 


The taginfo expecting a "static/fix url" for the project file, so it can be any url - as you like.
so If you don't want to store the project JSON file in the repo, no problem.

**Comment by kocio-pl** [comment]
>like this?
>old files | new files

Yes, this is what I meant.

> imho: than better if we move the taginfo JSON project file to the root. ( so no need a special folder )

This would be proper location IMO, if we want to store it in this repo. However I'm still not sure how do you want to run the script?

**Comment by HolgerJeromin** [comment]
> I guess this file will be created by Travis on next merge, isn't it?

We can not access travis files, so Travis can not generate the really used files. Ref #3012

**Comment by ImreSamu** [comment]
>However I'm still not sure how do you want to run the script?


* now I will generate for the current taginfo JSON file - with this pull request.
    * after accepted 
         * I try to add the  url to the https://github.com/taginfo/taginfo-projects/blob/master/project_list.txt

* every release 
    * re -RUN the python script 
    * and IF  the `git status` shows a difference ( and it seems correct ) 
          *  THEN  `git commit` the new taginfo JSON project file. 

*  AND every day the taginfo will process and show the info.

>IMO, if we want to store it in this repo.

what is the alternatives?

**Comment by HolgerJeromin** [comment]
> However I'm still not sure how do you want to run the script?

Imo a maintainer should generate the chicago preview png and this json file locally and commits both to this repo if there is a major change

**Comment by kocio-pl** [comment]
I said before that I don't like to make more manual steps if they can be done automatically - releasing is already quite complicated and checking the script output differences would be even worse. 

This script could be run from some other place and I guess even automatically checking the new release is possible, but as simple thing as running it every night for example would do too.

**Comment by ImreSamu** [comment]
@kocio-pl :  

I have finished.

Ready to final review, merge, adapt, delete, reject, modify ... as you like 

>This script could be run from some other place and I guess even automatically 

Sorry,  I can't help on this topic,

**Comment by kocio-pl** [comment]
Thanks. Unfortunately I can't do the merge here, since I'm not even a programmer. Some coder has to do the code review. Anybody willing?

File: scripts/taginfo-project/generate-taginfo-project-file.py [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+                    osmtype='area'
+            else:
+                # If no Geometry type - we try to guess the type.
+                if 'planet_osm_point' in ds_table.lower():
+                    osmtype='node'
+                elif 'planet_osm_polygon'  in ds_table.lower():
+                    osmtype='area'
+                elif 'planet_osm_line'  in ds_table.lower():
+                    osmtype='way'
+                elif 'planet_osm_ways'  in ds_table.lower():
```
Comment (pnorman) [inline scripts/taginfo-project/generate-taginfo-project-file.py]: We don't use the ways table, and can't without a schema change.

File: scripts/taginfo-project/generate-taginfo-project-file.py [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+                    osmtype='way'
+                else:
```
Comment (pnorman) [inline scripts/taginfo-project/generate-taginfo-project-file.py]: planet_osm_roads hasn't been processed

File: scripts/taginfo-project/generate-taginfo-project-file.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+
+
+
+#
+# Parsing "project.mml" file for the HSTORE keys (  tags-> )
+#
+
+with open( cartocss_project_file , 'r') as f:
+  newf = yaml.load(f.read())
+f.closed
```
Comment (pnorman) [inline scripts/taginfo-project/generate-taginfo-project-file.py]: Doesn't the with block make sure it's closed?

File: scripts/taginfo-project/README.md [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+* Be carefull with the SQL comments or use as a simple hack:  ` /* tags->'wetland' */ `
+* This code tested only on Ubuntu Linux
+* Check the result!
+
+### Valid examples for parsing
+* `tags ? 'wetland'"`
+* `tags @> '"generator:source"=>wind'`
+* `tags @> 'capital=>yes'"]`
+* `tags -> 'leaf_type'`
+* `tags -> 'wetland'`
```
Comment (pnorman) [inline scripts/taginfo-project/README.md]: There's already an example of this type one line up

File: scripts/taginfo-project/generate-taginfo-project-file.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+#  tags->'leaf_type'
+#  tags @> '"generator:source"=>wind'
+re_tags_one     = re.compile(r"[^a-zA-Z0-9_]tags[^'^)^\[^\]]*'.+?'")
+
+
+# ----------------------------------  Examples --------------------
+#  tags -> ARRAY['wheelchair',ramp:wheelchair']
+#  tags ?& ARRAY['wheelchair',ramp:wheelchair']
+#  tags ?| ARRAY['wheelchair',ramp:wheelchair']
+re_tags_array = re.compile(  r"[^a-zA-Z0-9_]tags\s*[@\?-][>&\|]\s*[aA][rR][rR][aA][yY]\[.+?\]"  )
```
Comment (pnorman) [inline scripts/taginfo-project/generate-taginfo-project-file.py]: Use [re.IGNORECASE](https://docs.python.org/3.1/library/re.html#re.IGNORECASE) as a flag for re.compile to simplify the expression

File: .travis.yml [hunk 7]
```diff
@@ -1,22 +1,32 @@
 language: node_js
 dist: trusty
-sudo: false
+sudo: required
```
Comment (pnorman) [inline .travis.yml]: This will have a significant impact on how long it takes us to get a Travis instance

File: .travis.yml [hunk 8]
```diff
@@ -1,22 +1,32 @@
...
 install:
+  - sudo apt-get install -y python3-yaml
   - npm install carto@$CARTO
   - mkdir -p data/world_boundaries data/simplified-land-polygons-complete-3857 data/ne_110m_admin_0_boundary_lines_land data/land-polygons-split-3857
   - touch data/simplified-land-polygons-complete-3857/simplified_land_polygons.shp data/ne_110m_admin_0_boundary_lines_land/ne_110m_admin_0_boundary_lines_land.shp data/land-polygons-split-3857/land_polygons.shp
 script:
   # We're using pipes in the checks, so fail if any part fails
   - set -o pipefail
+  # Check the Taginfo-project
+  - cd ./scripts/taginfo-project && ./generate-taginfo-project-file.py  && cd ../..
```
Comment (pnorman) [inline .travis.yml]: it should be callable as `scripts/taginfo-project/generate-taginfo-project-file.py`

File: .travis.yml [hunk 9]
```diff
@@ -1,22 +1,32 @@
...
+  - git diff taginfo-openstreetmap-carto.json
```
Comment (pnorman) [inline .travis.yml]: It's better to generate to a different file and diff the existing taginfo-openstreetmap-carto.json against the new one, or have the script write to stdout and use `<()`

**Comment by matthijsmelissen** [comment]
Thanks for proposing this, @ImreSamu.

The thing I'm worrying about is that this adds fairly complicated code that will break easily under many unsuspected code. Also it is quite a lot of code that achieves relatively little (it only parses the keys). Moreover, checking whether a key is used in the mml does not give any guarantees that the key is actually displayed.

So I doubt this is the right way to go. I'm again inclined now towards manually maintaining a list of interpreted tags (which might also serve as a way of specification for us).

**Comment by ImreSamu** [comment]
@matthijsmelissen 

Thank you for you feedback.

Imho: My proposal is not guaranteed to be optimal or perfect, but sufficient for my immediate goals. Just creating a "minimal" taginfo project file with a minimal "business value" for the OSM community.

**As I understand and accept that the maintainence problem is not optimal for my "toy code", 
so I will move my code to the separate repo/project, and I will close this pull requests in the next days.**

Now I am testing my code with other carto forks ( osmfr-cartocss, openstreetmap-carto-hiking-cz,  ... )

From the "minimal" taginfo project files I can create some new "business value" :
for example:

`osmfr-cartocss` extra keys ( compare to this project ) : 
```
CEMT aerodrome alt_name basin boat bus 
capacity:disabled conveying conveyor_dir craft crossing crossing:bollard ele:local footway fuel:lpg golf 
heritage importance incline information int_name is_capital level line location 
maritime medical motorboat mountain_pass name:fr network organic 
parking piste:difficulty piste:type post_office:type ramp:wheelchair ref:FR:LaPoste ref:INSEE room 
school:FR shelter_type short_name short_name:fr sloped_curb stars 
tactile_paving tidal type:RATP 
wall water wheelchair 
```

`carto-hiking-cz` extra keys:
```
kct_blue kct_green kct_red kct_yellow network official_name:cs osmc:symbol
````


>So I doubt this is the right way to go. I'm again inclined now towards manually maintaining a list of interpreted tags 
>(which might also serve as a way of specification for us).

My proposal is just a temporary ( quick&dirty )  solution. It is not perfect  and I accept that it has no 'business value' for the current project members. 
So I will be very happy If somebody create and maintain an improved version of `taginfo project file` with more business value.

**Pull request closed without merging**